src/toxcore/dht/server/hole_punching.rs

```diff
@@ -0,0 +1,363 @@
+
+use toxcore::crypto_core::*;
+use toxcore::dht::server::*;
+use toxcore::dht::packed_node::*;
+use toxcore::io_tokio::IoFuture;
+
+/// Interval in seconds for sending NatPingRequest
+pub const NAT_PING_PUNCHING_INTERVAL: u64 = 3;
+/// Interval in seconds to reset trying count of hole-punch,
+/// if connecttion to friends is not done for over 40 seconds
+/// retry to connect to friends using NatPingRequest
+pub const RESET_PUNCH_INTERVAL: u64 = 40;
+// Maximum number of ports to punch hole
+const MAX_PORTS_TO_PUNCH: u32 = 48;
+// Maximum trying count for hole punch
+// after RESET_PUNCH_INTERVAL seconds it is reset if connection is not done.
+const MAX_NORMAL_PUNCHING_TRIES: u32 = 5;
+/// Maximum clients number per friend
+pub const MAX_CLIENTS_PER_FRIEND: u32 = 8;
+
+/// Structure for hole punch
+pub struct HolePunching {
+    /// flag for hole punching is done or not.
+    pub is_punching_done: bool,
+    /// number of punching tries
+    pub num_punch_tries: u32,
+    /// last timestamp of receiving NatPingResponse packet
+    pub last_recv_ping_time: Instant,
+    /// last timestamp of sending NatPingRequest packet
+    pub last_send_ping_time: Instant,
+    /// last timestamp of trying hole punch
+    /// it is used to send NatPingRequest every 3 seconds
+    pub last_punching_time: Instant,
+    /// factor variable for guessing NAT port
+    pub first_punching_index: u32,
+    /// another factor variable for guessing NAT port
+    pub last_punching_index: u32,
+    /// holds ping_id used in NatPingRequest
+    /// multi NatPingRequest has this same ping_id
+    /// because every NatPingRequest receives NatPingResponse.
+    pub ping_id: u64,
+}
+
+impl HolePunching {
+    /// new object of HolePunching
+    pub fn new() -> Self {
+        HolePunching {
+            is_punching_done: true,
+            num_punch_tries: 0,
+            last_recv_ping_time: Instant::now(),
+            last_send_ping_time: Instant::now(),
+            last_punching_time: Instant::now(),
+            first_punching_index: 0,
+            last_punching_index: 0,
+            ping_id: HolePunching::new_ping_id(),
+        }
+    }
+
+    // get new ping id for NatPingRequest packet
+    fn new_ping_id() -> u64 {
+        loop {
+            let ping_id = random_u64();
+            if ping_id != 0 {
+                return ping_id;
+            }
+        }
+    }
+
+    /// send NatPingRequest and if condition is true, do hole punch
+    pub fn try_nat_punch(&mut self, server: &Server, friend_pk: PublicKey, addrs: Vec<SocketAddr>,
+                         nat_ping_req_interval: Duration) -> IoFuture<()> {
+        if !self.is_punching_done &&
+            self.last_punching_time.elapsed() >= nat_ping_req_interval &&
+            self.last_recv_ping_time.elapsed() <= nat_ping_req_interval * 2 {
+                let ip = match HolePunching::get_major_ip(&addrs, MAX_CLIENTS_PER_FRIEND / 2) {
+                    None => return Box::new(future::ok(())),
```

Comment:
do nothing because... why?

Reply:
Because we consider it is not behind NAT, so don't do hole punching.

Reply:
Then please comment it in the code